eth/63 fast synchronization algorithm

karalabe opened the pull request:
This PR aggregates a lot of small modifications to `core`, `trie`, `eth` and other packages to collectively implement the eth/63 fast synchronization algorithm. In short, `geth --fast`.

### Algorithm

The goal of the the fast sync algorithm is to exchange processing power for bandwidth usage. *Instead of processing the entire block-chain one link at a time, and replay all transactions that ever happened in history, fast syncing downloads the transaction receipts along the blocks, and pulls an entire recent state database.* This allows a fast synced node to still retain its status an an archive node containing all historical data for user queries (and thus not influence the network's health in general), but at the same time to reassemble a recent network state at a fraction of the time it would take full block processing.

An outline of the fast sync algorithm would be:
 * Similarly to classical sync, download the block headers and bodies that make up the blockchain
 * Similarly to classical sync, verify the header chain's consistency (POW, total difficulty, etc)
 * Instead of processing the blocks, download the transaction receipts as defined by the header
 * Store the downloaded blockchain, along with the receipt chain, enabling all historical queries
 * When the chain reaches a recent enough state (head - 1024 blocks), pause for state sync:
   * Retrieve the entire Merkel Patricia state trie defined by the root hash of the pivot point
   * For every account found in the trie, retrieve it's contract code and internal storage state trie
 * Upon successful trie download, mark the pivot point (head - 1024 blocks) as the current head
 * Import all remaining blocks (1024) by fully processing them as in the classical sync

### Analysis

By downloading and verifying the entire header chain, we can guarantee with all the security of the classical sync, that the hashes (receipts, state tries, etc) contained within the headers are valid. Based on those hashes, we can confidently download transaction receipts and the entire state trie afterwards. Additionally, by placing the pivoting point (where fast sync switches to block processing) a bit below the current head (1024 blocks), we can ensure that even larger chain reorganizations can be handled without the need of a new sync (as we have all the state going that many blocks back).

### Caveats

The historical block-processing based synchronization mechanism has two (approximately similarly costing) bottlenecks: *transaction processing* and *PoW verification*. The baseline fast sync algorithm successfully circumvents the transaction processing, skipping the need to iterate over every single state the system ever was in. However, verifying the proof of work associated with each header is still a notably CPU intensive operation.

However, we can notice an interesting phenomenon during header verification. With a negligible probability of error, we can still guarantee the validity of the chain, only by verifying every `K`-th header, instead of each and every one. By selecting a single header at random out of every `K` headers to verify, we guarantee the validity of an `N`-length chain with the probability of `(1/K)^(N/K)` (i.e. we have `1/K` chance to spot a forgery in `K` blocks, a verification that's repeated `N/K` times).

Let's define the *negligible probability* `Pn` as the probability of obtaining a 256 bit SHA3 collision (i.e. the hash Ethereum is built upon): `1/2^128`. To honor the Ethereum security requirements, we need to choose the minimum chain length `N` (below which we veriy every header) and maximum `K` verification batch size such as `(1/K)^(N/K) <= Pn` holds. Calculating this for various `{N, K}` pairs is pretty straighforward, a simple and lenient solution being http://play.golang.org/p/B-8sX_6Dq0.

| N    | K   |     | N    | K   |     | N    | K   |     | N    | K   |
| ---- | --- | --- | ---- | --- | --- | ---- | --- | --- | ---- | --- |
| 1024 |  43 |     | 1792 |  91 |     | 2560 | 143 |     | 3328 | 198 |
| 1152 |  51 |     | 1920 |  99 |     | 2688 | 152 |     | 3456 | 207 |
| 1280 |  58 |     | 2048 | 108 |     | 2816 | 161 |     | 3584 | 217 |
| 1408 |  66 |     | 2176 | 116 |     | 2944 | 170 |     | 3712 | 226 |
| 1536 |  74 |     | 2304 | 128 |     | 3072 | 179 |     | 3840 | 236 |
| 1664 |  82 |     | 2432 | 134 |     | 3200 | 189 |     | 3968 | 246 |

The above table should be interpreted in such a way, that if we verify every `K`-th header, after `N` headers the probability of a forgery is smaller than the probability of an attacker producing a SHA3 collision. It also means, that if a forgery is indeed detected, the last `N` headers should be discarded as not safe enough. Any `{N, K}` pair may be chosen from the above table, and to keep the numbers reasonably looking, we chose `N=2048, K=100`. This will be fine tuned later after being able to observe network bandwidth/latency effects and possibly behavior on more CPU limited devices.

Using this caveat however would mean, that the pivot point can be considered secure only after `N` headers have been imported after the pivot itself. To prove the pivot safe faster, we stop the "gapped verificatios" `X` headers before the pivot point, and verify every single header onward, including an additioanl `X` headers post-pivot before accepting the pivot's state. Given the above `N` and `K` numbers, we chose `X=24` as a safe number.

With this caveat calculated, the fast sync should be modified so that up to the `pivoting point - X`, only every `K=100`-th header should be verified (**at random**), after which all headers up to `pivot point + X` should be fully verified before starting state database downloading. **Note: if a sync fails due to header verification the last `N` headers must be discarded as they cannot be trusted enough.**

### Weakness

Blockchain protocols in general (i.e. Bitcoin, Ethereum, and the others) are susceptible to [Sybil attacks](https://en.wikipedia.org/wiki/Sybil_attack), where an attacker tries to completely isolate a node from the rest of the network, making it believe a false truth as to what the state of the real network is. This permits the attacker to spend certain funds in both the real network and this "fake bubble". However, the attacker can only maintain this state as long as it's feeding new valid blocks it itself is forging; and to successfully shadow the real network, it needs to do this with a chain height and difficulty close to the real network. In short, to pull off a successful Sybil attack, the attacker needs to match the network's hash rate, so it's a very expensive attack.

Compared to the classical Sybil attack, fast sync provides such an attacker with an extra ability, that of feeding a node a view of the network that's not only different from the real network, but also that might go around the EVM mechanics. The Ethereum protocol only validates state root hashes by processing all the transactions against the previous state root. But by skipping the transaction processing, we cannot prove that the state root contained within the fast sync pivot point is valid or not, so as long as an attacker can maintain a fake blockchain that's on par with the real network, it could create an invalid view of the network's state.

To avoid opening up nodes to this extra attacker ability, fast sync (beside being solely opt-in) will only ever run during an initial sync (i.e. when the node's own blockchain is empty). After a node managed to successfully sync with the network, fast sync is forever disabled. This way anybody can quickly catch up with the network, but after the node caught up, the extra attack vector is plugged in. This feature permits users to safely use the fast sync flag (`--fast`), without having to worry about potential state root attacks happening to them in the future. As an additional safety feature, if a fast sync fails close to or after the random pivot point, fast sync is disabled as a safety precaution and the node reverts to full, block-processing based synchronization.

### Performance

To benchmark the performance of the new algorithm, four separate tests were run: full syncing from scrath on Frontier and Olympic, using both the classical sync as well as the new sync mechanism. In all scenarios there were two nodes running on a single machine: a seed node featuring a fully synced database, and a leech node with only the genesis block pulling the data. In all test scenarios the seed node had a fast-synced database (smaller, less disk contention) and both nodes were given 1GB database cache (`--cache=1024`).

The machine running the tests was a Zenbook Pro, Core i7 4720HQ, 12GB RAM, 256GB m.2 SSD, Ubuntu 15.04.

|       Dataset (blocks, states)        | Normal sync (time, db) | Fast sync (time, db) |
| ------------------------------------- | ------------------ | ------------------- |
| Frontier, 357677 blocks, 42.4K states | 12:21 mins, 1.6 GB | 2:49 mins, 235.2 MB |
| Olympic, 837869 blocks, 10.2M states  | 4:07:55 hours, 21 GB | 31:32 mins, 3.8 GB  |

The resulting databases contain the entire blockchain (all blocks, all uncles, all transactions), every transaction receipt and generated logs, and the entire state trie of the head 1024 blocks. This allows a fast synced node to act as a full archive node from all intents and purposes.

### Closing remarks

The fast sync algorithm requires the functionality defined by eth/63. Because of this, testing in the live network requires for at least a handful of discoverable peers to update their nodes to eth/63. On the same note, verifying that the implementation is truly correct will also entail waiting for the wider deployment of eth/63.

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Fri Oct 9 07:30:50 UTC 2015_

karalabe commented:
Just a mental note, my chain assembly functions do not push chain events into the mux. This should probably be something to discuss as to what - if anything - should be pushed. Another open ended question is how to incorporate the state download progress into `eth.syncing` (we have no means to know the number of states we need to pull... can we estimate it on the client side? it's something to still figure out).

Gustav-Simonsson commented:
Great write up, given that the summary of the PR is quite thorough and this whole change is non-trivial, I wanted to start by nitpicking a bit about the summary itself. After review it could be compiled into a nice small paper/report pdf :) 

### Pivot Point at 1024

Having a point like this seems reasonable, but the choice of 1024 is arbitrary and needs better motivation. For example, in https://blog.ethereum.org/2015/09/14/on-slow-and-fast-block-times/ Vitalik argues practical finality (assuming no attackers with hash power very close to 51%) with a 17s block time happens around 2 minutes. This would give about 8 blocks for Ethereum.

So a much smaller number should be OK to configure here for the pivot point if the reason for it is to avoid the final sync happening within threshold for probable chain reorgs. Also please consider naming this point to something like "reorg [security] threshold" to make it more descriptive.

### K-th header PoW verification

Seems reasonable. It states we have `1/K` chance to spot forgery in `K` blocks, but I think what is meant here is `1/K` being the probability of spotting a forgery in *any given block*. For a range of `K` blocks the probability should be close to 1. Please clarify this. Also the later `(1/K)^(N/K)` refers not to probability of spotting forgery or correctness, but the reverse: probability of an attacker getting away with forgery. 

Clarifying this will make it easier for readers unfamiliar with blockchains and also when we later on refer to this PR post-merge.

I'd also add a reference to [FIPS 202 section A.1](http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf) which is the official claim to SHA3's collision resistance - this can be good given [recent discussions around SHA-1's collision resistance](https://threatpost.com/practical-sha-1-collision-months-not-years-away/114979/) which after research ended up being less than initially thought.

The table of `{N, K}` pairs is not entirely correct.

For example, for `N = 1024, K = 44` we get (please excuse Erlang console syntax):
```
1> Pn = 1/math:pow(2,128).
2.938735877055719e-39
2> f(N), N = 1024.
1024
3>  f(K), K = 44.
44
4> math:pow((1/K), (N/K)) < Pn.
false
5> f(K), K = 43.
43
6> math:pow((1/K), (N/K)) < Pn.
true
7> math:pow((1/K), (N/K)).     
1.2608347620717529e-39
```
(K needs to be 43 to satisfy `(1/K)^(N/K) <= Pn`)

For `N = 2048, K = 113` it seems some error in the calculation has accumulated more, as we need `K = 108` to satisfy the property:
```
8> f(N), N = 2048.             
2048
9> f(K), K = 113.              
113
10> math:pow((1/K), (N/K)) < Pn.
false
11> f(K), K = 108.
108
12> math:pow((1/K), (N/K)) < Pn.
true
13> math:pow((1/K), (N/K)).     
2.7558821171772803e-39
```

Finally, the analysis concludes by selecting `N=2048, K=100` due to "keep numbers reasonably looking". This is a somewhat arbitrary argument and needs more motivation. These numbers should be fine to configure exactly to achieve a certain probability threshold, even if that results in numbers not falling in common ranges such as powers of two or round decimal numbers.

Gustav-Simonsson commented:
The time for normal sync on Olympic is missing.

karalabe commented:
Good catch with the rounding error, I'll have to create a tad better code for it. Maybe I could use float64 too and not need to do this hula hoop jumping with big.Float. Regarding the Olympic sync time, I know, I started running it ad ran out of disk space, so the whole thing crashed after 3 hours :P Will try to run it again tonight.

karalabe commented:
I've corrected the rounding issue in the `{N, K}` generator, and updated the description with the new code and the corrected values. The selected values in the PR as well within range, so no stress there.

The selection of 2048 was really because at the current blockchain size, it requires approximately 5K header verifications, whereas the extremities of the listed values already approach 6+K.

The selection of `K` itself is not that very relevant (+- a few, from the implementation's point of view), as it's just a maximum gap. The fast sync code does the verifications whenever importing a batch, and **always** verifies the last of a batch, so when importing 192 (default batch size), we're actually doing 2 random checks + the last always, so practically, `K` suggest 100, but we have more like 66. Also, as we're only processing one batch of headers at a time, if you verify less than the CPU cores available, you'll either way have to wait for it to complete, so raising it higher doesn't gain you much. But I agree that we should decide on **why** it is some arbitrary value :)

Lastly for the pivot point... I don't think it looses us much if we process 1K, but maybe on an embedded system it's more painful so I'm happy with reducing it, just let's figure out a reasonable value to reduce to.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -141,30 +141,146 @@ func NewBlockChain(chainDb ethdb.Database, pow pow.PoW, mux *event.TypeMux) (*Bl
+	fastTd := self.GetTd(self.currentFastBlock.Hash())
+
+	glog.V(logger.Info).Infof("Last header: #%d [%x…] TD=%v", self.currentHeader.Number, self.currentHeader.Hash().Bytes()[:4], headerTd)
+	glog.V(logger.Info).Infof("Fast block: #%d [%x…] TD=%v", self.currentFastBlock.Number(), self.currentFastBlock.Hash().Bytes()[:4], fastTd)
```
Minor nitpick; switch order so Last Block is logged immediately after Last Header, since they belong together, whereas Fast Block could be another block.

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -141,30 +141,146 @@ func NewBlockChain(chainDb ethdb.Database, pow pow.PoW, mux *event.TypeMux) (*Bl
+	return nil
+}
+
+// SetHead rewind the local chain to a new head entity. In the case of headers,
```
Can drop entity; head is descriptive enough.

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -141,30 +141,146 @@ func NewBlockChain(chainDb ethdb.Database, pow pow.PoW, mux *event.TypeMux) (*Bl
+
+// SetHead rewind the local chain to a new head entity. In the case of headers,
+// everything above the new head will be deleted and the new one set. In the case
+// of blocks though, the head may be further rewound if block bodies are missing
```
Please consistently use present tense in comments preceding code; here past and future tense is mixed in; "will be deleted" -> "is deleted" and "rewound" -> "rewinded"

karalabe replied:
There's no such word as rewinded :P

karalabe replied:
And there's also no past tense in there. "may be rewound is future" :P

Gustav-Simonsson replied:
https://en.wiktionary.org/wiki/rewinded

karalabe replied:
http://www.macmillandictionary.com/spellcheck/british/?q=rewinded vs
http://www.macmillandictionary.com/dictionary/british/rewound

Don't get me wrong, it is used, but afaik it's a simplified version and not the original. But if you can get me a better source than macmillan, I can change my mind :)

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -141,30 +141,146 @@ func NewBlockChain(chainDb ethdb.Database, pow pow.PoW, mux *event.TypeMux) (*Bl
 
-	for block := bc.currentBlock; block != nil && block.Hash() != head.Hash(); block = bc.GetBlock(block.ParentHash()) {
-		DeleteBlock(bc.chainDb, block.Hash())
+	// Figure out the highest known canonical assignment
```
assignment -> headers & blocks

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -141,30 +141,146 @@ func NewBlockChain(chainDb ethdb.Database, pow pow.PoW, mux *event.TypeMux) (*Bl
+// FastSyncCommitHead sets the current head block to the one defined by the hash
+// irrelevant what the chain contents were prior.
+func (self *BlockChain) FastSyncCommitHead(hash common.Hash) error {
+	// Make sure that both the block as well at it's state trie exists
```
as well as its

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -181,6 +297,17 @@ func (self *BlockChain) LastBlockHash() common.Hash {
 }
 
+// CurrentHeader retrieves the current head header of the canonical chain. The
+// header is retrieved from the chain manager's internal cache.
```
In this and the two following comments, please rename chain manager to it's shiny new `blockchain` name :)

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -231,19 +347,12 @@ func (bc *BlockChain) Reset() {
 // specified genesis state.
 func (bc *BlockChain) ResetWithGenesisBlock(genesis *types.Block) {
+	// Dump the entire block chain and purge the caches
+	bc.SetHead(0)
```
Should this be after getting the lock?

karalabe replied:
SetHead has its own lock.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -231,19 +347,12 @@ func (bc *BlockChain) Reset() {
-	bc.blockCache.Purge()
-	bc.futureBlocks.Purge()
-
 	// Prepare the genesis block and reinitialize the chain
```
OK it's not your comment, but please change reinitialize to English spelling :)

karalabe replied:
Eh, don't get all this English vs. English thing :P But, changed in nonetheless so that we don't have to debate it :D :P Will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -290,17 +400,30 @@ func (self *BlockChain) ExportN(w io.Writer, first uint64, last uint64) error {
+// insert injects a new head block into the current block chain. This method
+// assumes that the block is indeed a true head. It will also reset the head
+// header and the head fast sync block to this very same block to prevent them
+// from diverging on a different header chain.
```
What does "on a different header chain" mean? Please drop or clarify this dependent clause.

karalabe replied:
Changed to

> It will also reset the head header and the head fast sync block to this very same block to prevent them from pointing to a possible old canonical chain (i.e. side chain by now).

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
 
+// writeHeader writes a header into the local chain, given that its parent is
+// already known. If the total difficulty of the newly inserted header becomes
+// greater than the old known TD, the canonical chain is re-routed.
```
"old known" -> "current" is more accurate, since it's not old yet, and may not be depending on logic in this function.

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+// greater than the old known TD, the canonical chain is re-routed.
+//
+// Note: This method is not concurrent-safe with inserting blocks simultaneously
+// into the chain, as side effects caused by reorganizations cannot be emulated
```
reorganisations

karalabe replied:
Waaay too much time on your hands. Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+	td := new(big.Int).Add(header.Difficulty, ptd)
+
+	// Make sure no inconsistent state is leaked during insertion
+	self.mu.Lock()
```
Do the writes of TD and header need locking? If not this lock could be taken only if the TD check below is true.

karalabe replied:
Yes, they do. I've seen quite a lot of weird anomalies while testing and they turned out to be caused by other parts of the code seeing partial database writes. Of course the part in the `if` is much more prone to this as it's a long running op compared to the td/header writing, but they too can lead to errors due to partial views of the database. Additionally, we need to obtain the mutex already to access the current head (in the if's condition).

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+		for i := header.Number.Uint64() + 1; GetCanonicalHash(self.chainDb, i) != (common.Hash{}); i++ {
+			DeleteCanonicalHash(self.chainDb, i)
+		}
+		// Overwrite any stale canonical number assignments
```
"canonical number assignments" can probably be simplified to "headers"

karalabe replied:
Canonical numbers aren't the headers. The canonical numbering is an association of block numbers to block hashes. No headers are rewritten or even touched, we only update what we believe to be the current best chain progression.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+	return nil
+}
+
+// InsertHeaderChain will attempt to insert the given header chain in to the
```
attempts

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+}
+
+// InsertHeaderChain will attempt to insert the given header chain in to the
+// local chain, possibly creating a fork. If an error is returned,  it will
```
would use "reorg" instead of "fork" here since fork can imply a difference in consensus protocol.

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+
+	// Generate the list of headers that should be POW verified
+	verify := make([]bool, len(chain))
+	for i := 0; i < len(verify)/checkFreq; i++ {
```
Would it make sense to always verify at least one PoW, in case `len(chain) < checkFreq` ?

karalabe replied:
I'll delete this review comment to hide this code snipped from appearing as pending.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+	// Generate the list of headers that should be POW verified
+	verify := make([]bool, len(chain))
+	for i := 0; i < len(verify)/checkFreq; i++ {
+		index := i*checkFreq + rand.Intn(checkFreq)
```
Since the reason behind the randomness is to make it harder for potential attackers, let's initialise math/rand with a seed from crypto/rand on startup of the node. This will make it significantly harder for attackers, and still allows this code to use math/rand API.

Here is the code todo this in ethash's GPU miner branch: https://github.com/Gustav-Simonsson/ethash/blob/go_gpu_miner/ethash_opencl.go#L361

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+		}
+		verify[index] = true
+	}
+	verify[len(verify)-1] = true // Last should always be verified to avoid junk
```
Ah, nevermind above comment then :)

karalabe replied:
I'll delete this review comment to hide this code snipped from appearing as pending.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+			}
+		}
+	}
+	// Start as many worker threads as goroutines allowed
```
Btw, did you try to spawn less than GOMAXPROCS to see if the max is indeed optimal? Don't know too much about Go scheduling but if the node is also processing other things it could be a lower number is optimal. Would be interesting to benchmark this on a full sync.

karalabe replied:
That would be a bit too much of a micro optimisation for me. Even if you benchmark that N is better than M, it will depend on you machine, the capacity of individual cores, ill also depend on what the machine itself as well as the node is doing in between. Finally, even if you manage to figure out a number, any future commit might make that number completely off. IMHO we should give Go a fighting chance to run everything concurrently, but leave it to its own scheduler to refrain itself if need be.

Gustav-Simonsson replied:
Makes sense, thanks for clarifying.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+			checkPow := verify[index]
+
+			var err error
+			if index == 0 {
```
Should not this header too be verified with it's parent?

karalabe replied:
Ah sorry, probably the name of the method is a bit weird. Both methods do a full verification, with the difference that for one I specifically provide the parent and for the other the parent is retrieved from the database. If the `index == 0`, then I need to reach out to the database since I don't have the parent header myself. On the other hand, i `index > 0`, then there's no need to do the expensive database lookup, if I know anyway what the parent is.

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -511,6 +630,333 @@ const (
+			atomic.AddInt32(&stats.processed, 1)
+		}
+	}
+	// Start as many worker threads as goroutines allowed
```
This one can be extracted into a helper function since it's equivalent to the one used in `InsertHeaderChain `

karalabe replied:
I'll think about it a bit more. I'm reluctant to split it out doe to making the code harder to understand, but I'll do it if I can find a clean enough way.

Gustav-Simonsson commented on core/state/sync.go:
```diff
@@ -0,0 +1,71 @@
+	"github.com/ethereum/go-ethereum/trie"
+)
+
+// StateSync is the main state  synchronisation scheduler, which provides yet the
```
double space

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/state/sync.go:
```diff
@@ -0,0 +1,71 @@
+
+// NewStateSync create a new state trie download scheduler.
+func NewStateSync(root common.Hash, database ethdb.Database) *StateSync {
+	// Pre-declare the result syncer t
```
t?

karalabe replied:
Eh, just some comment I forgot to finish while writing the code. Though I think it's clean enough without it so deleted the comment altogether.

Gustav-Simonsson commented on core/types/receipt.go:
```diff
@@ -27,89 +26,116 @@ import (
-
-func (self *Receipt) RlpEncode() []byte {
-	bytes, err := rlp.EncodeToBytes(self)
+// RlpEncode implements common.RlpEncode required for SHA derivation.
```
SHA3

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/types/receipt.go:
```diff
@@ -27,89 +26,116 @@ import (
+}
+
+// ReceiptForStorage is a wrapper around a Receipt that flattens and parses the
+// entire content of a receipt, opposed to only the consensus fields originally.
```
as opposed

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/types/receipt.go:
```diff
@@ -27,89 +26,116 @@ import (
 
-func (self *Receipt) String() string {
-	return fmt.Sprintf("receipt{med=%x cgas=%v bloom=%x logs=%v}", self.PostState, self.CumulativeGasUsed, self.Bloom, self.logs)
+// DecodeRLP implements rlp.Decoder, and loads the consensus fields of a receipt
```
both consensus and implementation

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/types/receipt.go:
```diff
@@ -27,89 +26,116 @@ import (
 
-func (self Receipts) RlpEncode() []byte {
-	bytes, err := rlp.EncodeToBytes(self)
+// RlpEncode implements common.RlpEncode required for SHA derivation.
```
SHA3

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on core/vm/log.go:
```diff
@@ -25,42 +25,47 @@ import (
-	TxIndex   uint
-	BlockHash common.Hash
-	Index     uint
+	// Derived fields (don't reorder!)
```
Why can't they be reordered?

karalabe replied:
Because `LogForStorage` is an automatic RLP encoder, which encodes the fields in order. So if you change the order, it will be incompatible with your current database contents. The reason I specifically added that comment is because the current field order is quite weird (block fields interrupted by transaction fields).

Gustav-Simonsson replied:
OK, makes sense. Could you please add this to the comment.

Gustav-Simonsson commented on core/vm/log.go:
```diff
@@ -25,42 +25,47 @@ import (
 type Logs []*Log
 
+// LogForStorage is a wrapper around a Log that flattens and parses the entire
+// content of a log, opposed to only the consensus fields originally (by hiding
```
as opposed

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/modes.go:
```diff
@@ -0,0 +1,26 @@
+type SyncMode int
+
+const (
+	FullSync  SyncMode = iota // Synchronise the entire block-chain history from full blocks
```
blockchain or block chain! not block-chain. :)

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/modes.go:
```diff
@@ -0,0 +1,26 @@
+
+const (
+	FullSync  SyncMode = iota // Synchronise the entire block-chain history from full blocks
+	FastSync                  // Quikcly download the headers, full sync only at the chain head
```
Quickly

karalabe replied:
Done, will push in next commit. (Weird that my spell checker didn't catch this).

Gustav-Simonsson commented on eth/downloader/peer.go:
```diff
@@ -137,55 +158,89 @@ func (p *peer) Fetch(request *fetchRequest) error {
+
+// SetBlocksIdle sets the peer to idle, allowing it to execute new retrieval requests.
+// Its block retrieval allowance will also be updated either up- or downwards,
+// depending on whether the previous fetch completed in time or not.
```
whether + negation is unneeded, can drop "or not"

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/peer.go:
```diff
@@ -137,55 +158,89 @@ func (p *peer) Fetch(request *fetchRequest) error {
+
+// SetBodiesIdle sets the peer to idle, allowing it to execute new retrieval requests.
 // Its block body retrieval allowance will also be updated either up- or downwards,
 // depending on whether the previous fetch completed in time or not.
```
same here

Gustav-Simonsson commented on eth/downloader/peer.go:
```diff
@@ -310,26 +378,63 @@ func (ps *peerSet) AllPeers() []*peer {
-		for j := i + 1; j < len(list); j++ {
-			if atomic.LoadInt32(&list[i].rep) < atomic.LoadInt32(&list[j].rep) {
-				list[i], list[j] = list[j], list[i]
+	for i := 0; i < len(idle); i++ {
```
![](http://i.imgur.com/nguHh2A.jpg)
I guess it's fine in practice given realistic peer set sizes, but we could make it sexy by using the `Sort` interface to get `O(n*log(n))`

karalabe replied:
Yeah, I've also considered using the built in sort, but that would require implementing `Less`, `Len` and `Swap` for a custom alias of the `[]*peer` type, and it's really not worth it when the average node will have 10-20 connections max. The code complexity outweighs the performance gain. Imho.

Gustav-Simonsson replied:
Fair enough.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -23,25 +23,32 @@ import (
 
 var (
-	blockCacheLimit = 8 * MaxBlockFetch // Maximum number of blocks to cache before throttling the download
+	blockCacheLimit = 1024 // Maximum number of blocks to cache before throttling the download
```
Any specific reasoning behind this number? Would it make sense to have different for light / full / archive nodes?

karalabe replied:
*[Well, currently there's no notion of a full node, only archive and light. This PR started out as a special sync mechanism for introducing full nodes, but we figured that it would cripple the node quite significantly (can't retrieve past logs and transactions) and also would hurt the network health, so until we figure out a way to keep the network guaranteed healthy we retained all history for now.]*

I think the reason behind 1024 previously was that it's low enough not to be a memory burden, and high enough to not be importable too fast. Now, with fast syncing where import times are an order of magnitude faster, maybe it would make sense to raise this. However a final value could be only tuned with a live network to see the latency effects. What value would you suggest for fast syncs? (Note, the name blockCacheLimit is a bit off, it would better be called resultCacheLimit as the downloader pulls not only blocks any more, but receipts too).

Gustav-Simonsson replied:
Cool, thanks for explaining. I don't know what value would make most sense, I'll see if I can measure a difference by tweaking it when syncing between nodes on a local network. Could you please add a code comment with this reasoning? I know at least I will forget about it when reading this code again in a few months :)

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -49,35 +56,73 @@ type fetchRequest struct {
 	Time    time.Time           // Time when the request was made
 }
 
+// fetchResult is the assembly collecting partial results from potentially more
```
This sentence is a bit funky, I guess it should be either "assembly of partial results" or "is collecting partial results"

karalabe replied:
Yeah, that indeed sounds... interesting :D Don't know where my head was.

karalabe replied:
```
// fetchResult is a struct collecting partial results from data fetchers until
// all outstanding pieces complete and the result as a whole can be processed.
```

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -49,35 +56,73 @@ type fetchRequest struct {
 // queue represents hashes that are either need fetching or are being fetched
 type queue struct {
+	mode          SyncMode // Synchronisation mode to decide on the block parts to schedule for fetching
+	fastSyncPivot uint64   // Block number where the fast sync pivots into archive synchronisation mode
```
Perhaps call it `fastSyncThreshold`

karalabe replied:
Yes, originally I called it threshold, but wanted something shorter. Calling it `fastSyncThreshold` is a way too long name in Go naming conventions and simply `threshold` is not descriptive enough to figure what it's used for. If you really dislike it I'll change it, but I'd rather use something short and sweet.

Gustav-Simonsson replied:
Hehe, yeah I don't want to go full Java style on variable names, but sometimes I think Go's convention favours a bit too short names.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -49,35 +56,73 @@ type fetchRequest struct {
+	receiptDonePool  map[common.Hash]struct{}      // [eth/63] Set of the completed receipt fetches
 
-	pendPool map[string]*fetchRequest // Currently pending block retrieval operations
+	stateTaskIndex int                      // [eth/63] Counter indexing the added hashes to ensure prioritized retrieval order
```
prioritised

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -49,35 +56,73 @@ type fetchRequest struct {
+	stateDatabase   ethdb.Database   // [eth/63] Trie database to populate during state reassembly
+	stateScheduler  *state.StateSync // [eth/63] State trie synchronisation scheduler and integrator
+	stateProcessors int32            // [eth/63] Number of currently running state processors
+	stateSchedLock  sync.RWMutex     // [eth/63] Lock serializing access to the state scheduler
```
serialising

Gustav-Simonsson replied:
I guess it should be `stateSchedMu` for consistent naming with other mutexes.

karalabe replied:
1. Done, will push in next commit.
2. Actually, the code is split between Mu and Lock. If not for other reason, because I personally used Lock everywhere :stuck_out_tongue_winking_eye: . On a more serious tone though, I quite heavily oppose to leaking type information into the variable name. Calling it a `Mu(tex)` is a tad much imho.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -49,35 +56,73 @@ type fetchRequest struct {
+	stateSchedLock  sync.RWMutex     // [eth/63] Lock serializing access to the state scheduler
+
+	resultCache  []*fetchResult // Downloaded but not yet delivered fetch results
+	resultOffset uint64         // Offset of the first cached fetch result in the block-chain
```
blockchain or block chain

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -86,81 +131,141 @@ func (q *queue) Reset() {
+	q.stateTaskQueue.Reset()
+	q.statePendPool = make(map[string]*fetchRequest)
+	q.stateScheduler = nil
+
```
Should q.stateProcessors also be reset?

karalabe replied:
No, because those are tied to running go routines. Resetting it would cause it to go into negative when those stale worker threads actually exit. One thing we might consider is whether such a stale worker could have any adverse effect on a subsequent sync, but I don't see any meaningful probability of them influencing the next sync cycle.

Gustav-Simonsson replied:
OK, makes sense. Just wanted to make a sanity check from diffing struct fields and the ones being reset.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -86,81 +131,141 @@ func (q *queue) Reset() {
-// Size retrieves the number of blocks in the queue, returning separately for
-// pending and already downloaded.
-func (q *queue) Size() (int, int) {
+// InFlightBlocks retrieves whether there are block fetch requests currently in
```
It retrieves number of reqs in flight, even if there are none.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -86,81 +131,141 @@ func (q *queue) Reset() {
-func (q *queue) Has(hash common.Hash) bool {
+// ThrottleReceipts checks if the download should be throttled (active receipt
+// fetches exceed block cache).
+func (q *queue) ThrottleReceipts() bool {
```
It's name make it sound like it's doing something; perhaps call it `NeedToThrottleReceipts` or some such.

karalabe replied:
Changed to `ShouldThrotleXxx`. Will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -186,22 +291,17 @@ func (q *queue) Insert61(hashes []common.Hash, fifo bool) []common.Hash {
 	q.lock.Lock()
 	defer q.lock.Unlock()
 
 	// Insert all the headers prioritized by the contained block number
```
prioritised

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -186,22 +291,17 @@ func (q *queue) Insert61(hashes []common.Hash, fifo bool) []common.Hash {
-			glog.V(logger.Warn).Infof("Header #%d [%x] already scheduled", header.Number.Uint64(), hash[:4])
-			continue
-		}
 		// Make sure chain order is honored and preserved throughout
```
honoured. respect ~~the Empire~~ British spelling we use in the code base :)

karalabe replied:
![Queen](http://sd.keepcalm-o-matic.co.uk/i/keep-calm-and-long-live-the-queen-4.png)

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -210,96 +310,160 @@ func (q *queue) Insert(headers []*types.Header, from uint64) []*types.Header {
+		}
+		// Queue the header for content retrieval
+		q.blockTaskPool[hash] = header
+		q.blockTaskQueue.Push(header, -float32(header.Number.Uint64()))
```
Hmm, why float32?

karalabe replied:
It's required by [my priority queue implementation](https://godoc.org/gopkg.in/karalabe/cookiejar.v2/collections/prque). The idea was that computationally and memory wise it costs the same as an int, but gives a greater flexibility if you need to insert items into the queue after already having a bunch in there. Not sure it it's a good or bad decision, but it's here to stay for now :P

Gustav-Simonsson replied:
Ah, OK. I probably disagree with that design decision - floats are evil and should be avoided unless there are very good reasons for them :) But in this case it shouldn't be a problem given the number of queue items we'll see in practice.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -210,96 +310,160 @@ func (q *queue) Insert(headers []*types.Header, from uint64) []*types.Header {
-		return nil
-	}
-	if _, ok := q.pendPool[p.id]; ok {
+	// Short circuit if the peer's already downloading something (sanity check not
```
to not

karalabe replied:
Although both should be fine, I'm ok with changing the order to make it a tad simpler form. Will push in the next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -319,49 +483,86 @@ func (q *queue) Reserve61(p *peer, count int) *fetchRequest {
+// previously failed downloads. Beside the next batch of needed fetches, it also
+// returns a flag whether empty blocks were queued requiring processing.
+func (q *queue) ReserveBodies(p *peer, count int) (*fetchRequest, bool, error) {
+	noop := func(header *types.Header) bool {
```
Perhaps call it `isNoop` since it does something and is used to see whether *another thing* will be a noop

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -319,49 +483,86 @@ func (q *queue) Reserve61(p *peer, count int) *fetchRequest {
-		if header.TxHash == types.DeriveSha(types.Transactions{}) && header.UncleHash == types.CalcUncleHash([]*types.Header{}) {
-			if err := q.enqueue("", types.NewBlockWithHeader(header)); err != nil {
-				return nil, false, errInvalidChain
+		// If we're the first to request this task, initialize the result container
```
initialise

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -319,49 +483,86 @@ func (q *queue) Reserve61(p *peer, count int) *fetchRequest {
+			}
+		}
+		// If this fetch task is a noop, skip this fetch operation
+		if noop(header) {
```
Since stuff is done if `noop` returns true it would be better to call it `skipFetch` or some such.

karalabe replied:
`skipFetch` doesn't really convey the reason it should be skipped. Noop imho better signals that you don't *need* to fetch because *no data* will be returned (hence the noop name).

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -319,49 +483,86 @@ func (q *queue) Reserve61(p *peer, count int) *fetchRequest {
 			continue
 		}
-		// If it's a content block, add to the body fetch request
+		// Otherwise if not a known unknown block, add to the retrieve list
```
known unknown?

karalabe replied:
known unknown-block. As in we know that the peer does not have that block.

Rephrased to "// Otherwise unless the peer is known not to have the data, add to the retrieve list"

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -370,81 +571,155 @@ func (q *queue) Reserve(p *peer, count int) (*fetchRequest, bool, error) {
+}
+
+// Revoke cancels all pending requests belonging to a given peer. This method is
+// meant to be called during a peer drop to quickly reassign owned data fetches
```
what does "owned data fetches" refer to here? and why are they reassigned? should not all fetches from a peer be dropped if the peer is dropped? are the reassigned fetches outgoing fetches from this node that have not been responded to yet?

karalabe replied:
It means, that whenever the downloader needs to pull some data, it assigns that piece of missing data (its hash) to a peer to avoid requesting the same data from somebody else too. Now, if we drop a peer, we need to return all missing data assigned to it to the pool so that others might fulfill them.

Gustav-Simonsson replied:
OK, makes sense.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -370,81 +571,155 @@ func (q *queue) Reserve(p *peer, count int) (*fetchRequest, bool, error) {
+}
+
+// ExpireBodies checks for in flight block body requests that exceeded a timeout
+// allowance, canceling them and returning the responsible peers for penalization.
```
penalisation (also for the two following functions)

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -473,60 +757,94 @@ func (q *queue) Deliver61(id string, blocks []*types.Block) (err error) {
+	case len(errs) == 1 && (errs[0] == errInvalidChain || errs[0] == errInvalidBlock):
+		return errs[0]
+
+	case len(errs) == len(request.Headers):
```
Could multiple errors of types not signifying stale delivery sum up to length of req headers?

karalabe replied:
Originally the errs array was collecting various types of errors, but by now all other errors have been handled separately (and checked for specifically in the above `case`s), so from this line onward `errs` can only contain errors related to getting data not requested. If all the data pieces were flagged as not requested, it's stale.

Gustav-Simonsson replied:
OK, would be great to put a short comment describing this above this case, as matching first on specific errors and then on length is a bit non-trivial.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -534,11 +852,8 @@ func (q *queue) Deliver(id string, txLists [][]*types.Transaction, uncleLists []
+	case len(errs) == 1 && (errs[0] == errInvalidChain || errs[0] == errInvalidBody || errs[0] == errInvalidReceipt):
+		return errs[0]
 
 	case len(errs) == len(request.Headers):
```
Same as above question.

karalabe replied:
Same as above answer :D

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -548,29 +863,101 @@ func (q *queue) Deliver(id string, txLists [][]*types.Transaction, uncleLists []
+	stateReqTimer.UpdateSince(request.Time)
+	delete(q.statePendPool, id)
+
+	// If no data was retrieved, mark them as unavailable for the origin peer
```
mark it

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/downloader/queue.go:
```diff
@@ -548,29 +863,101 @@ func (q *queue) Deliver(id string, txLists [][]*types.Transaction, uncleLists []
+		hash := common.BytesToHash(crypto.Sha3(blob))
+		if _, ok := request.Hashes[hash]; !ok {
+			errs = append(errs, fmt.Errorf("non-requested state data %x", hash))
+			continue
```
Should a peer be penalised for returning non-requested data?

karalabe replied:
It is. https://github.com/karalabe/go-ethereum/blob/fast-sync-rebase/eth/downloader/downloader.go#L1391

Gustav-Simonsson commented on eth/handler.go:
```diff
@@ -100,11 +97,15 @@ func NewProtocolManager(networkId int, mux *event.TypeMux, txpool txPool, pow po
+		if fastSync && version < eth63 {
+			continue
+		}
+		// Compatible, initialize the sub-protocol
```
initialise

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on eth/handler.go:
```diff
@@ -473,6 +461,28 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 		return p.SendBlockBodiesRLP(bodies)
 
+	case p.version >= eth62 && msg.Code == BlockBodiesMsg:
+		// A batch of block bodies arrived to one of our previous requests
```
Is it possible this code is executed if a BlockBodiesMsg is received which is not the result of previous requests?

karalabe replied:
We only request a single message of a specific kind from a peer at any time. One body fetch, one header, one receipt, one node data. The reply is assumed to be for that request. There's of course nothing stopping a peer from sending out of bound data, but that will just get ignored.

Note, there's one exception. There's a very tiny probability that the fetcher and downloader request blocks the same time from the same peer when transitioning from "doing a big sync" to "staying in sync", but the fetcher will explicitly cherry pick it's own requests and leave all others to the downloader, so this does not bother them.

Gustav-Simonsson replied:
OK, reason I was asking was just to improve the code comment - it reads as if the message will always be in response to node's request. Actually the comment can probably be removed, since the logic for ignoring unrequested messages is in the downloader.

Gustav-Simonsson commented on eth/handler.go:
```diff
@@ -500,6 +510,17 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 		return p.SendNodeData(data)
 
+	case p.version >= eth63 && msg.Code == NodeDataMsg:
+		// A batch of node state data arrived to one of our previous requests
```
Same here, it could get this message even if not requested right?

karalabe replied:
Yes, but then it will be simply ignored, so there's nut much harm.

Gustav-Simonsson commented on trie/sync.go:
```diff
@@ -0,0 +1,285 @@
+	object *node       // Target node to populate with retrieved data (hashnode originally)
+
+	parents []*request // Parent state nodes referencing this entry (notify all upon completion)
+	depth   int        // Depth level within the trie the node is located to prioritize DFS
```
prioritise

karalabe replied:
Done, will push in next commit.

Gustav-Simonsson commented on trie/sync.go:
```diff
@@ -0,0 +1,285 @@
+	"gopkg.in/karalabe/cookiejar.v2/collections/prque"
+)
+
+// request represents a scheduled or already in-flight state retrieval request.
```
Why are requests here scheduled? Can not each go routine handling requests/responses in the wire protocol simply read/write data using normal queries? This seems to introduce another queueing layer and I'm not understanding why it's needed.

karalabe replied:
Data retrievals in the `downloader/queue` is a simple flat list of hashes: "gimme this; here ya' go". The `request`s here form a trie themselves, and reassemble the trie in memory, writing out pieces as they get completed. The `downloader/queue` makes sure that required data is retrieved, handles concurrent retrievals from multiple peers, timeouts, retries, ignore lists, etc. This piece of code maintains a prioritized ordering of "future" requests that should be executed, an order that changes at every state insertion. The `downloader/queue` cannot reprioritize its fetch order as it has no idea which stte data is more important than the other... you need to interpret the data for that. So the downloader queue maintains only a very small list of currently active state downloads to handle any retrieval failures and retries, and the sync.trie will maintain the real queue that always returns the most urgent nodes to fetch.

Gustav-Simonsson commented:
@karalabe Cool, yeah agreed the exact values are not critical, but if we can get them as good as possible it'll save some processing in nodes, could be significant for some embedded / light systems.

![](http://i.imgur.com/SPzky.gif)

karalabe commented:
![Grammar](http://static2.fjcdn.com/comments/I+know+everybody+already+knows+but+it+s+a+statement+not+_8e96b8bf9daabfa8ffb83761f5698c2c.jpg)

@Gustav-Simonsson I've corrected many of the issues you raised, please take a look at the remaining ones and my comments to them.

karalabe commented:
Note to self, needed before completing this PR: singleton fast sync (mark in database), sync status updates (move into downloader fully).

karalabe commented:
@obscuren I've rebased on top of the mipmap-inclusive develop. If you have the time, please check that I didn't miss anything (last commit in this PR). Sadly git it leaking modifications with the rebase so I had to patch up a few things that the rebase lost.

karalabe commented:
I'll squash all the commits after ca96683 when reviewers give the thumbs up.

alexvandesande commented:
👏🏻

Gustav-Simonsson commented:
aside from two small comments and questions (see gitter chat) :+1:

Gustav-Simonsson commented on core/blockchain.go:
```diff
@@ -504,6 +632,337 @@ const (
+	// Generate the list of headers that should be POW verified
+	verify := make([]bool, len(chain))
+	for i := 0; i < len(verify)/checkFreq; i++ {
+		index := i*checkFreq + self.rand.Intn(checkFreq)
```
Actually this makes it a bit more predictable, as it guarantees there is one and only one PoW verification every `checkFreq` block. A better way would be to make it fully random by generating `len(chain)/checkFreq` indexes randomly in the range `0 to len(chain)`

karalabe replied:
Hmm, though then we wouldn't be able to guarantee that every K batch is at least minimally verified, so I'm not sure about it.

karalabe commented:
@Gustav-Simonsson @obscuren Squashed all the review and discussion commits into one, ready for merge from my part.

Gustav-Simonsson commented:
:+1: to merge. the comment about random K is not a blocker. though would be nice to improve later on, perhaps in combination with tweaks of other fast sync params.

jamesray1 commented:
From the initial comment:
> This allows a fast synced node to still retain its status an an archive node containing all historical data for user queries

This allows a fast synced node to contain all historical data for user queries (like a classically synced node, and thus not influence the network's health in general)...

yihaient commented:
Can you please explain how to do this for a beginner? I am unable to download 100% of the blocks and I have a working MAcbook Pro that is brand new and the latest Mist/Ethereum Wallet Desktop App. I do not understand please help! thanks

yihaient commented:
@karalabe @Gustav-Simonsson Problem Above ^^^^^

jjtny1 commented:
@karalabe I have a question on what you mean by 
"This allows a fast synced node to still retain its status an an archive node containing all historical data for user queries (and thus not influence the network's health in general)".
How can a node that only has block headers and a receipt chain function as an archive node and have all historic data? You are only receiving the state for the latest Merkle Patricia Trie. How are you getting historical data?

ivica7 commented:
@jjtny1 that would be my question too. I thought geth is downloading the state in parallel to downloading the block headers, hence it pulls some historical states on the way, but it's not downloading it completely and the complete state download happens only at the pivot point(???)

In https://github.com/ethereum/go-ethereum/issues/15001#issuecomment-370732526 @karalabe is talking about the difficulties because of morphing state during the fast sync.

helinwang commented:
> But by skipping the transaction processing, we cannot prove that the state root contained within the fast sync pivot point is valid or not, so as long as an attacker can maintain a fake blockchain that's on par with the real network, it could create an invalid view of the network's state.

Isn't state root of the pivot point valid if the normal sync from the pivot point to the tip (1024 blocks) gives the same state root hash with the tip?

lyh168 commented:
Hi I read this  article had some question, if i set fast sync, now my chaindata(include states)  is may 40G, but 1 year later, my chaindata may be large again, because the states trash history have more again, have some Solutions to  delete states trash history in operation? thks
